packager/packager.cc

```diff
@@ -479,34 +517,196 @@ ast::UnresolvedConstantLit *verifyConstant(core::MutableContext ctx, core::NameR
+    vector<pair<core::NameRef, uint16_t>> curPkg;
+    bool foundTestNS = false;
+
+    static constexpr uint16_t SKIP_BOUND_VAL = 0;
+
+public:
+    PackageNamespaces(core::Context ctx, const PackageInfoImpl &filePkg, bool isTestFile)
+        : packages(ctx.state.packageDB().packages()), filePkg(filePkg), begin(0), end(packages.size()),
+          isTestFile(isTestFile), filePkgIdx(findPackageIndex(ctx, filePkg)) {
+        ENFORCE(packages.size() < numeric_limits<uint16_t>::max());
+    }
+
+    int depth() const {
+        return nameParts.size();
+    }
+
+    const vector<core::NameRef> &currentConstantName() const {
+        return nameParts;
+    }
+
+    core::NameRef packageForNamespace(core::Context ctx) const {
+        if (curPkg.empty()) {
+            return core::NameRef::noName();
+        }
+        return curPkg.back().first;
+    }
+
+    bool onPackagePath(core::Context ctx) {
+        if (begin <= filePkgIdx && filePkgIdx < end) {
+            return true;
+        }
+        if (!curPkg.empty() && curPkg.back().first == filePkg.mangledName()) {
+            return true;
+        }
+        return false;
+    }
+
+    void pushName(core::Context ctx, core::NameRef name) {
+        if (skips > 0) {
+            skips++;
+            return;
+        }
+        uint16_t idx = bounds.size();
+
+        if (isTestFile && idx == 0 && !foundTestNS) {
+            if (isPrimaryTestNamespace(name)) {
+                foundTestNS = true;
+                return;
+            } else if (!isTestNamespace(ctx, name)) {
+                bounds.emplace_back(begin, end);
+                nameParts.emplace_back(name);
+                begin = end = 0;
+                return;
+            }
+        }
+
+        if (idx > 0 && end - begin == 1 && packages[begin] == filePkg.mangledName()) {
+            curPkg.emplace_back(packages[begin], SKIP_BOUND_VAL);
+            skips++;
+            return;
+        }
+
+        bounds.emplace_back(begin, end);
+        nameParts.emplace_back(name);
+        auto lb = std::lower_bound(packages.begin() + begin, packages.begin() + end, nameParts,
+                                   [ctx](auto pkgNr, auto &nameParts) -> bool {
+                                       return core::packages::PackageInfo::lexCmp(
+                                           ctx.state.packageDB().getPackageInfo(pkgNr).fullName(), nameParts);
+                                   });
+        auto ub =
+            std::upper_bound(lb, packages.begin() + end, LexNext(nameParts), [ctx](auto &next, auto pkgNr) -> bool {
+                return next < ctx.state.packageDB().getPackageInfo(pkgNr).fullName();
+            });
+
+        begin = lb - packages.begin();
+        end = ub - packages.begin();
+
+        if (begin != end) {
+            auto &pkgInfo = ctx.state.packageDB().getPackageInfo(*lb);
+            ENFORCE(pkgInfo.exists());
+            if (bounds.size() == pkgInfo.fullName().size()) {
```

Comment:
Why not compare against `nameParts.size()`?

Reply:
Fixed. Remnant of older code that left nameParts in the tree walk.